core/src/main/java/org/infinispan/persistence/marshallers/MapMarshaller.java

```diff
@@ -0,0 +1,66 @@
+package org.infinispan.persistence.marshallers;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+import org.infinispan.protostream.MessageMarshaller;
+import org.infinispan.util.KeyValuePair;
+
+public class MapMarshaller implements MessageMarshaller<Map> {
+
+   private final Class<? extends Map> clazz;
+
+   MapMarshaller(Class<? extends Map> clazz) {
+      this.clazz = clazz;
+   }
+
+   @Override
+   public Map readFrom(ProtoStreamReader reader) throws IOException {
+      String implementation = reader.readString("implementation");
+      return reader
+            .readCollection("entry", new ArrayList<>(), KeyValuePair.class)
+            .stream()
+            .collect(Collectors.toMap(KeyValuePair::getKey, KeyValuePair::getValue, (o, n) -> n, createMap(implementation)));
+   }
+
+   private Supplier<Map> createMap(String implementation) {
+      return () -> {
+         try {
+            return implementation == null ? new HashMap() : (Map) Class.forName(implementation).newInstance();
+         } catch (ClassNotFoundException | InstantiationException | IllegalAccessException e) {
+            throw new IllegalStateException(e);
+         }
+      };
+   }
+
+   @Override
+   public void writeTo(ProtoStreamWriter writer, Map map) throws IOException {
+      if (!map.isEmpty()) {
+         List<KeyValuePair> kvp = ((Set<Map.Entry>) map.entrySet()).stream()
+               .map(e -> new KeyValuePair<>(e.getKey(), e.getValue()))
```

Comment:
Man this is looking even worse than before.... 👎 Wish there was a nice way to do this.

Reply:
@anistor When are you aiming to release 4.3.0? I'm mainly thinking of https://issues.jboss.org/browse/IPROTO-74